src/utils/utils_parallelize.c

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright(c) 2012-2021 Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "ocf/ocf.h"
+#include "../engine/cache_engine.h"
+#include "../engine/engine_common.h"
+#include "../ocf_request.h"
+#include "utils_parallelize.h"
+
+struct ocf_parallelize {
+	ocf_cache_t cache;
+	ocf_parallelize_handle_t handle;
+	ocf_parallelize_finish_t finish;
+	void *priv;
+
+	unsigned shards_cnt;
+	env_atomic remaining;
+	env_atomic error;
+
+	struct ocf_request *reqs[];
+};
+
+static int _ocf_parallelize_hndl(struct ocf_request *req)
+{
+	ocf_parallelize_t parallelize = req->priv;
+	ocf_parallelize_finish_t finish;
+	void *priv;
+	int error;
+
+	error = parallelize->handle(parallelize, parallelize->priv,
+			req->byte_position, parallelize->shards_cnt);
+
+	env_atomic_cmpxchg(&parallelize->error, 0, error);
+
+	if (env_atomic_dec_return(&parallelize->remaining))
```

Comment:
We cannot do that, because we need to call `finish()` to free resources (so we cannot skip `finish()`), but we also cannot call it before **all** shards finished their work, because shards are using those resources. We also have no means to cancel shards work, so the only available solution is to just wait.